civis/_utils.py

```diff
@@ -153,3 +173,22 @@ def read(self, size=-1):
         data = self.buf.read(bytes_to_read)
         self.bytes_read += len(data)
         return data
+
+
+class wait_for_retry_after_header(wait_base):
+    """Wait strategy that first looks for Retry-After header. If not
+        present it uses the fallback strategy as the wait param"""
+    def __init__(self, fallback):
+        self.fallback = fallback
+
+    def __call__(self, retry_state):
+        # retry_state is an instance of tenacity.RetryCallState.
+        # The .outcome property contains the result/exception
+        # that came from the underlying function.
+        result_headers = retry_state.outcome._result.headers
```

Comment:
We're not throwing exceptions so I don't think so unless I'm missing something? The retry logic is using status code alone to trigger a retry. The reason for this is the default requests exception handling doesnt seem to pass back any of the response information for requests to handle. At one point I had a raise_for_status call in there hoping that i could set the `retry` parameter to something like `lambda res: res.status_code in civis.civis.RETRY_CODES & StatusException`. That would never trigger though because res.status_code didnt exist in the retry object. 

Also to take a step back this is sort of unrelated to the cause of the retry. All this class is doing is overriding the wait parameter with the header value if it exists. Does that make sense?

Reply:
@skomanduri does that sound good?

Reply:
Wouldn't `retry_state.outcome` contain an exception if requests throws an exception?  The situation I'm worried about is where that happens and then trying to call `.headers` on it would throw its own exception and we would lose the requests exception.  I don't know if this happens, but want to either rule it out or handle it.

Reply:
I'm unclear on what type of exception you are envisioning. I thought the only exceptions that requests would implicitly return are ConnectionError, Timeout, and TooManyRedirects. I don't think any of these would come into play with the retry logic as it currently exists. Since we're purely using the status_code returned on the response object, that wait time class will only be evaluated once the retry logic has been entered afaik.

Is there another set of exceptions that I'm missing here?

That said, currently if an exception is thrown currently the retry is not triggered.

Reply:
> That said, currently if an exception is thrown currently the retry is not triggered.

Okay, that's the piece I was missing.  I thought if there was a ConnectionError it would still evaluate here, but if not then this would be fine.  Can you add a test for no retry on ConnectionError?

Reply:
Okay test is last on the test_utils.py file.

Reply:
That's what I had in mind for the test.  WFM!

Reply:
@skomanduri can i get the final approval so i can merge this?